### PR TITLE
Allow caret move to next part after short year input (T711074)

### PR DESCRIPTION
**Pull request by KsyushaM**

### CODE REVIEW

**Review by mrjohn0011 — DISMISSED** [review]

File: js/ui/date_box/ui.date_box.mask.js [hunk 1]
```diff
@@ -138,14 +138,17 @@ let DateBoxMask = DateBoxBase.inherit({
     },
 
     _searchNumber(char) {
-        let limits = this._getActivePartLimits(),
-            maxSearchLength = String(limits.max).length;
+        const limits = this._getActivePartLimits();
+        const maxLimitLength = String(limits.max).length;
```
Comment (churkin) [inline js/ui/date_box/ui.date_box.mask.js]: Why we can not use `const maxLimitLength = limits.max` ?

**Review by mrjohn0011 — DISMISSED** [review]

File: js/ui/date_box/ui.date_box.mask.js [hunk 2]
```diff
@@ -138,14 +138,17 @@ let DateBoxMask = DateBoxBase.inherit({
     },
 
     _searchNumber(char) {
-        let limits = this._getActivePartLimits(),
-            maxSearchLength = String(limits.max).length;
+        const limits = this._getActivePartLimits();
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.mask.js]:
```suggestion
        const { max } = this._getActivePartLimits();
```

File: js/ui/date_box/ui.date_box.mask.js (hunk 1 above)
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.mask.js]:
```suggestion
        const maxLimitLength = String(max).length;
```

File: js/ui/date_box/ui.date_box.mask.js [hunk 3]
```diff
@@ -138,14 +138,17 @@ let DateBoxMask = DateBoxBase.inherit({
...
+        const formatLength = this._getActivePartProp("pattern").length;
 
-        this._searchValue = (this._searchValue + char).substr(-maxSearchLength);
+        this._searchValue = (this._searchValue + char).substr(-maxLimitLength);
 
         this._setActivePartValue(this._searchValue);
 
         if(this.option("advanceCaret")) {
+            const isShortFormat = formatLength === 1;
+            const maxSearchLength = (isShortFormat ? maxLimitLength : Math.min(formatLength, maxLimitLength));
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.mask.js]:
```suggestion
            const maxSearchLength = isShortFormat ? maxLimitLength : Math.min(formatLength, maxLimitLength);
```

File: js/ui/date_box/ui.date_box.mask.js [hunk 4]
```diff
@@ -116,8 +116,8 @@ let DateBoxMask = DateBoxBase.inherit({
             return this._formatPattern;
         }
 
-        var format = this._strategy.getDisplayFormat(this.option("displayFormat")),
-            isLDMLPattern = typeof format === "string" && (format.indexOf("0") >= 0 || format.indexOf("#") >= 0);
+        const format = this._strategy.getDisplayFormat(this.option("displayFormat"));
+        const isLDMLPattern = typeof format === "string" && !dateLocalization._getPatternByFormat(format);
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.mask.js]:
```suggestion
        const isLDMLPattern = isString(format) && !dateLocalization._getPatternByFormat(format);
```

Comment (DokaRus) [reply js/ui/date_box/ui.date_box.mask.js]: import from `utils/type`

**Review by mrjohn0011 — APPROVED** [review]

**Review by DokaRus — APPROVED** [review]

**Pull request merged**